### PR TITLE
feat(osqp_interface): add interface to give csc matrix directly to enable warm start for computation time improvement.

### PR DESCRIPTION
**Pull request by takayuki5168**
## Related Issue(required)

https://github.com/autowarefoundation/autoware.universe/issues/347

## Description(required)

Add a new interface in `osqp_interface` to enable warm-start which improves the computation time of the optimization. Details are below.

- added initializer of osqp_interface with csc matrix, and matrix update functions with csc matrix
- added a function to validate the size of P, q, A, l, u.
- fixed the existing test and added test for new functions

NOTE: we do not have to care about the existing applications which use osqp_interface since the existing member functions' behavior of osqp_interface will not change.

## Review Procedure(required)

run test

## Pre-Review Checklist for the PR Author

**PR Author should check the checkboxes below when creating the PR.**

- [x] Read [commit-guidelines][commit-guidelines]
- [x] Assign PR to reviewer

If you are adding new package following items are required:

- [ ] Documentation with description of the package is available
- [ ] A sample launch file and parameter file are available if the package contains executable nodes

## Checklist for the PR Reviewer

**Reviewers should check the checkboxes below before approval.**

- [x] Commits are properly organized and messages are according to the guideline
- [x] PR title describes the changes

## Post-Review Checklist for the PR Author

**PR Author should check the checkboxes below before merging.**

- [ ] All open points are addressed and tracked via issues or tickets

## CI Checks

- **Build and test for PR / build-and-test-pr**: Required to pass before the merge.
- **Build and test for PR / clang-tidy-pr**: NOT required to pass before the merge. It is up to the reviewer(s). Found false positives? See the [guidelines][clang-tidy-guidelines].
- **Check spelling**: NOT required to pass before the merge. It is up to the reviewer(s). See [here][spell-check-dict] if you want to add some words to the spell check dictionary.

[commit-guidelines]: https://www.conventionalcommits.org/en/v1.0.0/
[spell-check-dict]: https://github.com/tier4/autoware-spell-check-dict#how-to-contribute

### CODE REVIEW

File: common/osqp_interface/include/osqp_interface/osqp_interface.hpp [hunk 1]
```diff
@@ -57,7 +57,7 @@ class OSQP_INTERFACE_PUBLIC OSQPInterface
   int64_t m_exitflag;
 
   // Runs the solver on the stored problem.
-  std::tuple<std::vector<float64_t>, std::vector<float64_t>, int64_t, int64_t> solve();
+  std::tuple<std::vector<float64_t>, std::vector<float64_t>, int64_t, int64_t, int64_t> solve();
```
Comment (takayuki5168) [inline common/osqp_interface/include/osqp_interface/osqp_interface.hpp:60]: add iteration, which I guess an important value to the application

File: common/osqp_interface/src/osqp_interface.cpp [hunk 2]
```diff
@@ -184,11 +203,43 @@ int64_t OSQPInterface::initializeProblem(
...
+  if (A.rows() != static_cast<int>(l.size())) {
+    ss << "A.rows() and l.size() are not the same. A.rows() = " << A.rows()
+       << ", l.size() = " << l.size();
+    throw std::invalid_argument(ss.str());
+  }
+  if (A.rows() != static_cast<int>(u.size())) {
+    ss << "A.rows() and u.size() are not the same. A.rows() = " << A.rows()
+       << ", u.size() = " << u.size();
+    throw std::invalid_argument(ss.str());
+  }
```
Comment (takayuki5168) [inline common/osqp_interface/src/osqp_interface.cpp:232]: check if rols and cols of matrix, vector among P, q, A u, l have no conflict. (Once this checking function was in autoware, but dropped during .universe porting.)

File: common/osqp_interface/test/test_osqp_interface.cpp [hunk 3]
```diff
@@ -23,67 +23,109 @@
...
   auto check_result =
-    [](const std::tuple<std::vector<float64_t>, std::vector<float64_t>, int, int> & result) {
-      EXPECT_EQ(std::get<2>(result), 1);
-      EXPECT_EQ(std::get<3>(result), 1);
-      ASSERT_EQ(std::get<0>(result).size(), size_t(2));
-      ASSERT_EQ(std::get<1>(result).size(), size_t(2));
-      EXPECT_DOUBLE_EQ(std::get<0>(result)[0], 0.3);
-      EXPECT_DOUBLE_EQ(std::get<0>(result)[1], 0.7);
-      EXPECT_DOUBLE_EQ(std::get<1>(result)[0], -2.9);
-      EXPECT_NEAR(std::get<1>(result)[1], 0.2, 1e-6);
```
Comment (takayuki5168) [inline common/osqp_interface/test/test_osqp_interface.cpp:38]: image

Comment (takayuki5168) [reply common/osqp_interface/test/test_osqp_interface.cpp:38]: image
https://github.com/autowarefoundation/autoware.universe/blob/tier4/proposal/common/osqp_interface/src/osqp_interface.cpp/#L243
https://github.com/tier4/autoware.universe/tree/feature/osqp-add-functions

**Comment by takayuki5168** [comment]
I created the following PR with awf/'s branch wrong. So I created another PR with tier4's one.
https://github.com/autowarefoundation/autoware.universe/pull/346

File: common/osqp_interface/test/test_osqp_interface.cpp (hunk 3 above)
Comment (takayuki5168) [reply common/osqp_interface/test/test_osqp_interface.cpp:38]: @TakaHoribe Thank you for the fix! I confirmed the code and the test passed. :pray: 
@rej55 It seems that implementation to extract lagrange multiplier was wrong.

**Review by TakaHoribe — DISMISSED** [review]
LGTM. If it is ok for @rej55, please merge it.

(There are too many outputs in `optimize()` function. It'd be better to create a struct or use `getXxx()` Api to access the iteration or dual variables. But it is not a scope of this PR. Think later.)

**Review by rej55 — APPROVED** [review]
Thank you for checking the test code problem.
LGTM.

**Review by TakaHoribe — APPROVED** [review]

**Pull request merged**